[mdoc] Platform/Owner tagging

joelmartinez opened the pull request:
You can now maintain documentation for a library that is supported on multiple platforms in a single set of documentation. Types and members will be tagged with information about what platform it is supported on. The slightly more general concept of "owner" was chosen because this feature will eventually support more than just platforms ... it will ideally also support things like being able to tag the documentation with what PCL profile it's supported in. 

The biggest change is that the docs are now stored in the “unified” form on disk. After an `mdoc update`,
the XML files will be stored in folders that lack the “dropped namespace” (previously, it was stored in “classic” form, or with the dropped namespace). This makes it easier to have multiple frameworks that share common namespaces (other than the dropped prefix). Running mdoc update with the appropriate `dropns` parameter  will reformat and move files into the now-correct folders. If there are conflicting type XMLs between the two assemblies that are being merged, the version from the first assembly on the `dropns` list will be given preference.

This pull request also slightly changes the format for dropping namespaces on both the update and assemble
commands from previous versions.

To add owner information … just add `=ownerName` to the assembly list. So in the example below, Assembly1 is attributed to "Framework1", while Assembly2 is attributed to "Framework2":

    mdoc update Assembly1.dll=Framework1 Assembly2.dll=Framework2 -o en

Dropping namespaces … the assembly from which the namespace is being dropped is now added to the list
of dropped namespaces. This makes it easier from an implementation standpoint to know which types and members should the resulting namespace be dropped from:

    mdoc update Assembly1.dll=Framework1 Assembly2.dll=Framework2 --dropns=Assembly1.dll=DroppedNS --dropns=Assembly2.dll=OtherDroppedNS

When assembling the classic doc set (ie. tree/zip) after a namespace has been dropped, you now specify the namespaces that have been dropped. Unified does not need any special command line formulation (it was reversed previously) for this particular feature (though you'd still have to pass in `ntypes` if the "unified" assembly has those special native types):

    mdoc assemble en/ --droppedns=DroppedNS --droppedns=OtherDroppedNS --out=out-docs

The corresponding changes to support the display of this new "owner" information has been added to monodoc and the XSL templates.

joelmartinez commented:
@jonpryor Happy new years :) wanted to follow up on this. Are there any other pieces of feedback, or issues with this pull request that I need to address?

jonpryor commented:
@joelmartinez As per recent mono email thread, we would like PRs to be a single (squashed) commit (for cleaner history, etc.). It's also troubling that the Merged Build failed (though I'm not sure why it failed).

joelmartinez commented:
@jonpryor oh, I kept the two commits separate because they were changes to different areas (mdoc and monodoc). But let me squash those. 

I believe the build failure was unrelated to these changes as the mdoc and monodoc test suites are passing (at least, for me locally). But those results have been since removed from jenkins ... once I push up these changes squashed, it should kick off another build, so let's see what's up then :)

akoeplinger commented:
@jonpryor @joelmartinez I interpreted that email in the way that we shouldn't have multiple commits in the PR doing "uninteresting" things like fixing typos, adjust formatting etc. Those should be combined into the original commit. IMO it's perfectly fine to use separate commits where it makes sense, i.e. per work unit as has been done here.

joelmartinez commented:
I can go either way with this :) But in any case, while I was looking at this again, I discovered another edge case ... I've added another test to reproduce it and am in the process of fixing it. I will submit another update once that's fixed.

directhex commented:
build

joelmartinez commented:
Ahh, thanks for triggering that rebuild. I will take a look at what's causing the failure.

joelmartinez commented:
Just a quick update ... this build failure has proven tricky to track down. I fixed the code I thought it was introducing this issue, however, I haven't been able to reproduce this error locally, and the error keeps occurring. So I've added some additional error handling/output to see if I can track down the issue more effectively.

Waiting for this build to run ... Stay tuned :)

akoeplinger commented:
@joelmartinez I can reproduce this by going into `./mcs/docs` and running `make clean && make`.

joelmartinez commented:
Ahh, finally ... got this sorted :) thanks for that tip @akoeplinger, it was really helpful. 

I've also gone ahead and squashed the two commits into one. The build should work, or at least, not fail due to mdoc/monodoc ... let's see how this goes.

joelmartinez commented:
Well it looks like the build failed, but if I'm reading the build output correctly, it doesn't have anything to do with mdoc or monodoc ... thoughts?

akoeplinger commented:
build

akoeplinger commented:
The build failed due to https://github.com/mono/mono/commit/bb9d05a5f5438423ec8b376e8d9668920388d063, so let's try another one.

joelmartinez commented:
@akoeplinger should I try to rebase from master or something?

akoeplinger commented:
@joelmartinez yeah , you can try that, though I think something is weird with the build workers atm :D

akoeplinger commented on mcs/tools/mdoc/Test/en.expected.importecmadoc/System/Action`1.xml:
```diff
@@ -21,13 +21,12 @@
-					<para> Represents the method that performs an action on the specified object.</para>
-				</summary>
+      <para> Represents the method that performs an action on the specified object.</para>
+    </summary>
```
@joelmartinez Are those whitespace changes in the en.expected.importecmadoc files intended? We talked about it via email a while back, looks like this is why the Jenkins build fails.

joelmartinez replied:
Is it? just out of curiosity could you point out where you see this in the build spew (in the interest of me learning how to interpret the results better)? 

That being said, I don't remember making this change explicitly ... I rebased my branch *after* that commit was merged into master that we talked about via email (the one with the xml formatter changed).

akoeplinger replied:
@joelmartinez it's actually quite easy, just grep for the first occurence of "Error 1" in the [build output](https://jenkins.mono-project.com/job/test-mono-pull-request/357/label=debian-amd64/consoleText).

> in theory, it should match whatever is in master ... no?

No, as in that case it wouldn't show up as a change here :)

joelmartinez replied:
oh geez ... duh, of course, had a momentary brainfart and didn't realize this was the diff against master :P 

Ahh I see, so then the build output changed when you ran the build again? I guess it was just a temporary glitch with the build server? cool ... in any case. let me rebase against master and make sure there aren't any errant XML changes.

joelmartinez commented:
@akoeplinger rebased to master, and double checked that the test that was failing due to the formatting issue you pointed out is not failing after the rebase, nor does it have that funky formatting. Honestly, not really sure as I didn't make any changes other than simply to rebase. Let's see if this build passes.

akoeplinger commented:
@joelmartinez I still see those formatting changes for en.expected.importedecmadoc included in the diff, are you sure you reverted them (simply rebasing won't work I guess since you squashed everything into a single commit before)?

joelmartinez commented:
Ok, this is really weird ... I don't understand why these tests are not failing locally. When I run `make check-monodocer-importecmadoc` locally, it works, and the diff is passing correctly.

*sigh* my apologies for the back and forth on this. I guess I will manually copy down those particular files that are failing in the build output and see how that works? I'm guessing something went wonky with my original rebase and maybe I reverted the XML changes we were talking about earlier ... but it still doesn't explain why it's not failing locally. Oh well, we'll figure it out :)

akoeplinger commented:
@joelmartinez fwiw, I pulled down this PR and ran `make check-monodocer-importecmadoc` and it failed for me locally, so something might be weird with your checkout. Try a `git clean -xdf` (will wipe any untracked changes) and then building.

joelmartinez commented:
![image](https://cloud.githubusercontent.com/assets/90380/5745056/f7306066-9bf3-11e4-8457-40abeb22b6b0.png)

akoeplinger commented:
Finally :smile:  :tada:

jonpryor commented on mcs/class/monodoc/Monodoc.Ecma/EcmaDesc.cs:
```diff
@@ -232,20 +232,27 @@ void ConstructCRef (StringBuilder sb)
-				foreach (var t in GenericTypeArguments) {
-					if (i > 0) {
-						sb.Append (",");
+				if (GenericTypeArguments.Any (a => a == null)) {
```
I don't like this, because it's not particularly clear why "if any argument is null, use the count".

It also seems duplicative; there's *already* a `EcmaDesc.GenericTypeArgumentsIsNumeric` property, which seems to have the same *intent* behind it; shouldn't `GenericTypeArgumentsIsNumeric` be used here, instead of `.Any()`?

Also, `GenericTypeArgumentsIsNumeric` should probably be updated to use `.Any()` instead of `.FirstOrDefault()`.

joelmartinez replied:
I didn't notice that property ... so I will definitely refactor to use it (along with the other few comments in this area), thanks for pointing it out.

As for the intent behind this ... it's because when building up the ecma URLs, if the "type" is available, we can print out URLs such as SomeClass<T,K> ... but if even one of those type arguments is null, then we have no way of printing out the correct signature, and so it needs to default back to the SomeClass`2 form of the signature.

jonpryor commented on mcs/class/monodoc/Monodoc.Ecma/EcmaDesc.cs:
```diff
@@ -232,20 +232,27 @@ void ConstructCRef (StringBuilder sb)
-					if (i > 0) {
-						sb.Append (",");
+				if (GenericTypeArguments.Any (a => a == null)) {
+					sb.AppendFormat ("`{0}", GenericTypeArguments.Count ());
```
Why `GenericTypeArguments.Count()` when there's a `GenericTypeArgumentsCount` property? Is there any reason to prefer one over the other?

This should probably use `GenericTypeArgumentsCount`.

jonpryor commented on mcs/class/monodoc/Monodoc.Ecma/EcmaDesc.cs:
```diff
@@ -232,20 +232,27 @@ void ConstructCRef (StringBuilder sb)
+				if (GenericTypeArguments.Any (a => a == null)) {
+					sb.AppendFormat ("`{0}", GenericTypeArguments.Count ());
+				}
+				else {
```
This `else` block could be avoided -- and the corresponding whitespace changes due to indenting this block -- if the above `if` block used `continue`:

    if (GenericTypeArgumentsIsNumeric) {
        sb.AppendFormat ("`{0}", GenericTypeArgumentsCount);
        continue;
    }
    ...

jonpryor commented on mcs/class/monodoc/Monodoc/providers/EcmaDoc.cs:
```diff
@@ -31,7 +31,8 @@ internal class EcmaDoc
 		                                              Dictionary<string, XElement> nsSummaries,
 		                                              Func<XElement, string> indexGenerator = null,
-		                                              IEcmaProviderFileSource fileSource = null)
+		                                              IEcmaProviderFileSource fileSource = null,
```
`PopulateTreeFromIndexFile()` taking **8** parameters is a code smell; this is probably a sign that some refactoring is needed, e.g. turning this method into a class and each method parameter into a class property.

It's also odd that EcmaDoc has only static members but is not a static class.

Fortunately, the EcmaDoc type is internal, but still; this is crazy.

joelmartinez replied:
Yes, as with the mutable static state, this is also another thing that I hesitantly added to avoid a big refactor. The only reason I went forward with it is because the parameters I was adding were optional, and really only useful when being called specifically from mdoc ... so it wouldn't break any other clients that might be using this API. 

But you're right that it's a code smell ... if any cause comes up again to modify this code, I will certainly refactor rather than continue adding to the parameter list.

jonpryor commented on mcs/class/monodoc/Monodoc/providers/EcmaDoc.cs:
```diff
@@ -578,16 +602,19 @@ internal static string MakeOperatorSignature (XElement member, out string member
-			var docs = typeDoc.Root.Element("Docs");
-			var summary = docs.Element("summary") ?? new XElement("summary");
-			var remarks = docs.Element("remarks") ?? new XElement("remarks");
+			string name = typeDoc.Root.Attribute ("Name").Value;
```
Please avoid whitespace changes. This patch is huge enough as it is.

joelmartinez replied:
:/ I know, sorry ... it just seemed such an egregious break from the coding standards that had just made it through to master. Is it verboten to do whitespace changes ever? Would it be better to do have a "whitespace only" pull request to fix things like this? or better to just leave it as is until that code gets rewritten entirely at some point?

jonpryor replied:
The problem with whitespace changes is that they pollute history. You can't "just" do `git blame` to see when the line was added, because now the line was changed for purely whitespace reasons.

Thus, instead of:

    git blame file.cs

it's:

    git blame file.cs
    # get commit of line change; see it's just a whitespace change
    git blame COMMIT^ file.cs
    # repeat

and if `file` was ever MOVED, then `git blame COMMIT^ file.cs` fails because the path is wrong, and this is MADNESS.

Whitespace changes are evil if you ever need to dig through code history to figure out *why* a change was done.

> Is it verboten to do whitespace changes ever?

Let's go with "yes". You can do whitespace changes *in addition to other changes which impact the same line*, e.g. if a variable name changes and thus the line would change *anyway*, then additional whitespace changes to that line would be fine. (Though because of the complications in viewing history, renaming a variable "just because" is also something to be discouraged.)

joelmartinez replied:
Ok ... I will be very much more careful with this in the future. Thanks for the explanation :)

jonpryor commented on mcs/tools/mdoc/Mono.Documentation/assembler.cs:
```diff
@@ -76,22 +82,30 @@ public override void Run (IEnumerable<string> args)
 
 		List<Provider> list = new List<Provider> ();
-		EcmaProvider ecma = null;
+		List<EcmaProvider> ecma = null;
```
I don't understand why this is now a `List<EcmaProvider>`, particularly since only one provider will ever be added to it -- when `ecma == null`! `ecma.Add()` is only performed once; there's no need for this to be a list.

joelmartinez replied:
Ahh, you know what ... this is an artifact from an earlier implementation that I was going for. I eventually changed how I was going to do multiple runs (because I needed to be able to do everything, for both classic and unified assemblies in one invocation, but it's much easier to keep the logic separate while it's running ... so I just run it twice internally for each set of assemblies).

I can change this and will resubmit

jonpryor commented on mcs/tools/mdoc/Mono.Documentation/monodocer.cs:
```diff
@@ -165,35 +169,73 @@ class MDocUpdater : MDocCommand
 
-	public static string droppedNamespace = string.Empty;
+	public static DroppedAssemblyInfo.List droppedAssemblyList = new DroppedAssemblyInfo.List ();
+	public static string currentDroppedNamespace = string.Empty;
```
Could we *please* not increase the amount of mutable static state?

`mdoc` is known to be slow; a "reasonable" way to improve this would be to make it multithreaded, which in theory sounds like it should be easy because each Type should be (reasonably) independent, so just fire off a thread per Type and be done, right?

But shared static state makes this impossible; we need to *decrease* the amount of static mutable state, not increase it! (Decreasing static mutable state also makes thinking about changes easier, as there are fewer side effects to consider.)

joelmartinez replied:
This is very good feedback ... I only added the static state originally due to lack of time to do a bigger refactor, and this data needed to be referenced from a number of parts of the codebase. That being said, I didn't add this lightly, I'm aware of the limitations that come with adding static mutable state ... in fact, one of the things that I'm planning on doing is running mdoc through Xamarin's new profiler with an eye towards making it generally faster. Multithreading it is the obvious candidate to do that (other than general code/algorithm improvement) and this was first on my list to refactor out.

Is it cool if we leave the mutable state refactor until a later date? or do you want it resolved before merging this code in?

jonpryor replied:
It can wait.

Overall, I have two issues with this patch:

1. It's *gigantic*, which makes reviewing it hard.
2. One of the reasons why it's gigantic is because of pointless whitespace changes (mentioned elsewhere, but not all whitespace changes were commented upon).

The whitespace changes need to die.

That aside, is there some way to perhaps split this up into smaller changes?

joelmartinez replied:
For this particular thing, no ... all of this functionality is for maintaining documentation from two frameworks in a single set of XML docs ... since there is much shared code in this case. I'd originally kept the changes to both mdoc and monodoc in separate commits to try to break it up in that way, but I've since squashed those into one per early feedback on this PR.

But yes, generally I prefer to make smaller PRs when possible and avoid big "releases" like this ... this just happened to be a big job with lots of internal changes required :/

jonpryor commented on mcs/tools/mdoc/Mono.Documentation/monodocer.cs:
```diff
@@ -258,15 +300,23 @@ public override void Run (IEnumerable<string> args)
+
+						droppedAssemblies.Add (ass);
+						droppedNamespaces.Add (pref);
+						currentDroppedNamespace = pref;
```
...I mean, we're in `Run()`, which is an *instance* method, yet we go off and set a *static* variable? Why?

Yes, the codebase is in bad shape, but we need to clean things like this up, not make it worse!

jonpryor commented on mcs/tools/mdoc/Mono.Documentation/monodocer.cs:
```diff
@@ -274,51 +324,136 @@ public override void Run (IEnumerable<string> args)
-			assemblyResolver.AddSearchDirectory (dir);
+		ProcessAssemblyOwners (ref assemblyNames);
+
+		// check to see if there are two dropns entries
```
Why do we care? What's so special about there being 1 vs. 2 "droppedAssemblies" entries? What happens if there's 3? 4?

joelmartinez replied:
Well ... the purpose of the code is, specifically, to support a combined ios and mac  set of documentation, so the true purpose was always for two sets of assemblies from which you're dropping a namespace. I added support for one to make writing the unit tests easier and is how I developed the bulk of the functionality initially (tdd) ... but any more than that delves into territory that I didn't test, and didn't need to test, so if there are more than two it just assumes you're not dropping a namespace.

joelmartinez commented:
@jonpryor I've made the requested changes to remove the unnecessary `List<EcmaProvider>`, and also to reuse the `GenericTypeArgumentsCount` and `GenericTypeArgumentsIsNumeric` properties.

joelmartinez commented:
@akoeplinger looks like the build failed again due to unrelated issues? 
```
make[10]: *** [run-test-lib] Error 1
make[10]: Target `run-test-local' not remade because of errors.
make[10]: Leaving directory `/var/lib/jenkins/workspace/test-mono-pull-request/label/debian-amd64/mcs/class/System'
```
Does the build need to be triggered again?

akoeplinger commented:
@joelmartinez master is in flux lately, wait a bit for it to settle down :)

joelmartinez commented:
ahh ok, no worries :)

joelmartinez commented:
@akoeplinger ... has the master branch stabilized? any chance you could re-trigger the build to see if it passes checks? :)

akoeplinger commented:
@joelmartinez let's just try it :) the ARM builds will still fail however. Btw you should be able to trigger a build as well since everyone in the Xamarin org should be whitelisted.

akoeplinger commented:
build

migueldeicaza commented:
It looks like the build failed for reasons unrelated to this patch.

Jonathan, any updates on this?

joelmartinez commented:
FYI, @migueldeicaza, I'm making the requested changes for the icons and layout. I've already implemented the new icons, and am looking at the layout changes. I'll ping this thread when I push a new version that includes those :)

joelmartinez commented:
quick FYI on the status of this PR ... we are going to change the way we are deploying API docs, as such please don't accept this PR for a few weeks :P I should be done and ready for this to be merged sometime in early April

jonpryor commented:
@joelmartinez: Then while you're doing that, could you please split this up into smaller parts so that it's not a 6400+ line patch? :-)

joelmartinez commented:
@jonpryor ... I've split the PR into two separate commits; one for mdoc and one for monodoc.

@migueldeicaza ... this latest commit resolves the issue you encountered with the `GetAppearance` member on `UIActivityIndicatorView`. I tested it after refreshing the merged repository and verified that that member correctly contained both the classic and unified signatures (in addition to the two non-generic versions that were added in the update); and also that subsequent re-runs will not produce duplicate nodes.

Please note that if/when we commit to moving to this version of mdoc, the `mdoc update` invocation will be slightly different. I've included all the details in that email that I'd sent previously, but you can ping me for any additional details.

jonpryor commented:
> This feature allows you to assign framework or platform ownership to types
and individual members.

What does this *mean*? What is a "platform owner"? An email address? URI? "Just" a name?

Examples?

Why do we *care*? What's the use case? What uses it?

Is "owner" even the right *name*? For example, the commit adds `owner-ios-32.png` and `owner-mac-32.png` images. This is presumably to show in the documentation which "owners" are supported.

When describing the resulting web page, is that the wording you'd use? "The method String.ToString() is owned by iOS and Mac."?

No. Not it isn't.

Instead, look at the mdoc commit message:

> Owner … just add `=owner` to the assembly list:
> 
>     mdoc update Assembly1=Framework1 Assembly2=Framework2 -o en

The example uses "Framework1", *not* owner.

Thus, the entire "owner" concept is misleading; it should be "framework", "frameworkName", "frameworkIdentifier", etc.

> Additionally, there is also a bugfix included for EcmaDesc's ToEcmaCref method; 

That should probably be a separate commit/PR.

joelmartinez commented:
@jonpryor the "magic type" is a holdover from what ultimately became the "native types" that came along with the unified version of the iOS APIs (ie. `nint`, `nfloat`, etc.) ... this was just implemented before we started referring to them as native types. I can change how this is referenced internally to match the parameter that kicks it off (`-ntypes`)

joelmartinez commented:
@jonpryor in response to the comment about the [naming of 'owner'](https://github.com/mono/mono/pull/1460#issuecomment-82485688) ... I'll explain the use case, and my thinking behind the name. Totally open to suggestions on how we should proceed, but at least the discussion can have more information :)

So the core requirement I was given for this feature was to be able to store Mac and iOS API documentation in the same set of XML documents. So originally, what I eventually called `owner` was going to be `platform` based on the requirements alone. The reason I decided to generalize the term; yes, the ambiguity was, in some ways, on purpose; was because I saw that this could be used to support a feature that we've seen from support feedback as a necessity ... richer versioning information for API docs. Similar to what they have on MSDN:
![image](https://cloud.githubusercontent.com/assets/90380/6695419/71a61194-ccb7-11e4-969e-1d8c8d554e90.png)
To give you an example of where this comes up, people regularly come across [`System.Web` on the Android API docs](http://androidapi.xamarin.com/?link=N:System.Web) and email in to support asking why this doesn't work on Android. Of course the reason is that we simply have the mono BCL docs on there, but not the version that's supported on Android.

I reasoned that perhaps in the future, we could run this process against all of the relevant incarnations of the BCL and be able to tell the user that a given method was supported on, say: ios, android, mac, windows, pclX, pclY, pclZ. And so, I chose a more general name, `owner` (as in, "this API is owned by both MonoTouch and Xamarin.iOS". This particular phrasing came up in conversation during discussions.

But for the time being, all I needed to support was ios and mac, and so I left it as `owner` in hopes that we could leverage the feature in the future to expand the semantic value of our docs with all this additional information.

So given that information ... do you think that simply renaming this to `framework` would suffice?

jonpryor commented:
>  I chose a more general name, `owner` (as in, "this API is owned by both MonoTouch and Xamarin.iOS".

My problem is that I can't imagine actually using that phrasing, ever. Maybe that's just me.

I can imagine saying "This API is _provided by_ Xamarin.iOS and Xamarin.Android", or "supported in", or "exists", and probably any number of synonyms for those, but "owned by" is just...weird.

Furthermore, "Supported in" is the phrasing that MSDN uses:

> Supported in: 3.0, 2.0, 1.0
> Supported in: Windows Phone 8.1
> Supported in: Windows Phone Silverlight 8.1
> ...

Based on that, I'd go with "supported in" as the phrasing.

jonpryor commented:
Then there's the previous comment about how to use/provide this information, e.g. the example in the original commit message:

     mdoc update Assembly1.dll=Framework1 Assembly2.dll=Framework2 -o en

It's...not quite right?

Let's use a specific example:

    mdoc update mscorlib.dll=Xamarin.iOS System.dll=Xamarin.Android

The problem is that we _know_ `mscorlib.dll` will also be provided in _both_ Xamarin.iOS and Xamarin.Android (and the desktop, and...), and the same is true for `System.dll`.

Does the above command line even make _sense_?

The _intent_, as I understand it, that the current `<Member/>`/etc. elements would have an `<Owner/>` child element, which would contain these "owner names", and specifying the owner name after the assembly would allow all types/members from that assembly to be _updated_ to have new `<Owner/>` elements.

Assuming that's true, why do you need to support providing multiple different "owners" at the same time? Why not just specify it _once_ and have it be used by _all_ assemblies?

    mdoc update --supported-in=Xamarin.iOS mscorlib.dll ...

Furthermore, and sorry for not actually checking, what happens for members that *were* supported in previous "owner" assemblies but are not anymore?

For example, between public releases we do work on `mscorlib.dll`, adding `Mono.Runtime.SomeMethod()`:

    mdoc update --supported-in=Xamarin.iOS mscorlib.dll...

At this point, we'll produce a `Runtime.xml` file in which the `SomeMethod` member has `<Owner>Xamarin.iOS</Owner>`.

Then, assume this was actually a mistake, and Xamarin.iOS *shouldn't* have `Runtime.SomeMethod()`, so it gets removed from the Xamarin.iOS profile, and we update the docs...

    mdoc update --supported-in=Xamarin.iOS mscorlib.dll...

Will the above command go through *every XML file and member* to *remove* `<Owner/>` elements that match `--supported-in` but *don't currently exist` in the reference assemblies?

If not, then I think it should.

joelmartinez commented:
@jonpryor, I've created a separate pull request that breaks out several of the unrelated modifications present in this pull request (other than in the sense that they were necessary to implement for *this* feature to work): https://github.com/mono/mono/pull/1641

@migueldeicaza, in particular, the commit that unblocks you is f75548e8790283d581648982f58d4ddf7277fb69

In the meantime, I will rebase this branch (ie. #1460) to include the changes from #1641). Should I just close this and start a new/fresh PR when that's done?

jonpryor commented:
@joelmartinez: It should be fine staying with this PR.

lewurm commented:
I'm not quite sure if #1641 covers all the changes now? @joelmartinez: can you either rebase or close this PR?

joelmartinez commented:
That particular PR was just a few bug fixes that I broke out ... however, you're right that the strategy has changed. This functionality will be provided by the changes coming in https://github.com/mono/mono/pull/2377 and beyond
